File: programs/autocrat/src/instructions/conditional_swap.rs
use amm::instructions::swap;

use super::*;

#[derive(AnchorSerialize, AnchorDeserialize, Debug, Clone)]
pub struct ConditionalSwapParams {
    pub market: Market,
    pub swap_type: SwapType,
    pub input_amount: u64,
    pub min_output_amount: u64,
}

#[derive(Accounts)]
pub struct ConditionalSwap<'info> {
    #[account(mut, has_one = amm_base_vault, has_one = amm_quote_vault)]
    pub futarchy_amm: Box<Account<'info, FutarchyAmm>>,
    #[account(mut, associated_token::mint = futarchy_amm.base_mint, associated_token::authority = futarchy_amm)]
    pub amm_base_vault: Account<'info, TokenAccount>,
    #[account(mut, associated_token::mint = futarchy_amm.quote_mint, associated_token::authority = futarchy_amm)]
    pub amm_quote_vault: Account<'info, TokenAccount>,
    #[account(mut, associated_token::mint = base_vault.conditional_token_mints[1], associated_token::authority = futarchy_amm)]
    pub amm_pass_base_vault: Account<'info, TokenAccount>,
    #[account(mut, associated_token::mint = quote_vault.conditional_token_mints[1], associated_token::authority = futarchy_amm)]
    pub amm_pass_quote_vault: Account<'info, TokenAccount>,
    #[account(mut, associated_token::mint = base_vault.conditional_token_mints[0], associated_token::authority = futarchy_amm)]
    pub amm_fail_base_vault: Account<'info, TokenAccount>,
    #[account(mut, associated_token::mint = quote_vault.conditional_token_mints[0], associated_token::authority = futarchy_amm)]
    pub amm_fail_quote_vault: Account<'info, TokenAccount>,
    #[account(mut)]
    pub user_input_account: Account<'info, TokenAccount>,
    #[account(mut)]
    pub user_output_account: Account<'info, TokenAccount>,
    #[account(mut)]
    pub base_vault: Box<Account<'info, conditional_vault::state::ConditionalVault>>,
    #[account(mut)]
    pub base_vault_underlying_token_account: Box<Account<'info, TokenAccount>>,
    #[account(mut)]
    pub quote_vault_underlying_token_account: Box<Account<'info, TokenAccount>>,
    #[account(mut)]
    pub quote_vault: Box<Account<'info, conditional_vault::state::ConditionalVault>>,
    pub token_program: Program<'info, Token>,
    pub trader: Signer<'info>,
    #[account(mut)]
    pub pass_base_mint: Box<Account<'info, Mint>>,
    #[account(mut)]
    pub fail_base_mint: Box<Account<'info, Mint>>,
    #[account(mut)]
    pub pass_quote_mint: Box<Account<'info, Mint>>,
    #[account(mut)]
    pub fail_quote_mint: Box<Account<'info, Mint>>,
    pub conditional_vault_program: Program<'info, ConditionalVaultProgram>,
    /// CHECK:
    pub vault_event_authority: UncheckedAccount<'info>,
    pub question: Account<'info, Question>,
}

impl ConditionalSwap<'_> {
    pub fn handle(ctx: Context<Self>, params: ConditionalSwapParams) -> Result<()> {
        let ConditionalSwapParams {
            market,
            swap_type,
            input_amount,
            min_output_amount,
        } = params;

        assert_ne!(market, Market::Spot);

        let output_amount =
            ctx.accounts
                .futarchy_amm
                .state
                .swap(input_amount, swap_type, market)?;

        require_gte!(output_amount, min_output_amount);

        // You need to transfer in before you can do merges of in
        // You need to do split of out before you can do transfers of out

        let amm_input_account = match (swap_type, market) {
            (SwapType::Buy, Market::Pass) => &ctx.accounts.amm_pass_quote_vault,
            (SwapType::Sell, Market::Pass) => &ctx.accounts.amm_pass_base_vault,
            (SwapType::Buy, Market::Fail) => &ctx.accounts.amm_fail_quote_vault,
            (SwapType::Sell, Market::Fail) => &ctx.accounts.amm_fail_base_vault,
            (_, Market::Spot) => unreachable!(),
        };

        require_gte!(ctx.accounts.user_input_account.amount, input_amount);

        token::transfer(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                token::Transfer {
                    from: ctx.accounts.user_input_account.to_account_info(),
                    to: amm_input_account.to_account_info(),
                    authority: ctx.accounts.trader.to_account_info(),
                },
            ),
            input_amount,
        )?;

        // We reload these to ensure that `quote_mergeable` and `base_mergeable` are accurate
        ctx.accounts.amm_pass_base_vault.reload()?;
        ctx.accounts.amm_pass_quote_vault.reload()?;
        ctx.accounts.amm_fail_base_vault.reload()?;
        ctx.accounts.amm_fail_quote_vault.reload()?;

        let signer_seeds = &[
            b"futarchy_amm".as_ref(),
            &[ctx.accounts.futarchy_amm.pda_bump],
        ];
        let signer = &[&signer_seeds[..]];

        let quote_cpi_context = CpiContext::new_with_signer(
            ctx.accounts.conditional_vault_program.to_account_info(),
            conditional_vault::cpi::accounts::InteractWithVault {
                question: ctx.accounts.question.to_account_info(),
                vault: ctx.accounts.quote_vault.to_account_info(),
                vault_underlying_token_account: ctx
                    .accounts
                    .quote_vault_underlying_token_account
                    .to_account_info(),
                authority: ctx.accounts.futarchy_amm.to_account_info(),
                user_underlying_token_account: ctx.accounts.amm_quote_vault.to_account_info(),
                event_authority: ctx.accounts.vault_event_authority.to_account_info(),
                program: ctx.accounts.conditional_vault_program.to_account_info(),
                token_program: ctx.accounts.token_program.to_account_info(),
            },
            signer,
        )
        .with_remaining_accounts(vec![
            ctx.accounts.fail_quote_mint.to_account_info(),
            ctx.accounts.pass_quote_mint.to_account_info(),
            ctx.accounts.amm_fail_quote_vault.to_account_info(),
            ctx.accounts.amm_pass_quote_vault.to_account_info(),
        ]);

        let amm_output_account = match (swap_type, market) {
            (SwapType::Buy, Market::Pass) => &ctx.accounts.amm_pass_base_vault,
            (SwapType::Sell, Market::Pass) => &ctx.accounts.amm_pass_quote_vault,
            (SwapType::Buy, Market::Fail) => &ctx.accounts.amm_fail_base_vault,
            (SwapType::Sell, Market::Fail) => &ctx.accounts.amm_fail_quote_vault,
            (_, Market::Spot) => unreachable!(),
        };

        // If the user is buying, we should have just received some quote to merge
        // If they're selling, we might need to split some quote
        match swap_type {
            SwapType::Buy => {
                let quote_mergeable = std::cmp::min(
                    ctx.accounts.amm_fail_quote_vault.amount,
                    ctx.accounts.amm_pass_quote_vault.amount,
                );

                if quote_mergeable > 0 {
                    conditional_vault::cpi::merge_tokens(quote_cpi_context, quote_mergeable)?
                }
            }
            SwapType::Sell => {
                let amount_to_split = output_amount.saturating_sub(amm_output_account.amount);
                if amount_to_split > 0 {
                    conditional_vault::cpi::split_tokens(
                        quote_cpi_context,
                        amount_to_split,
                    )?
                }
            }
        }

        let base_cpi_context = CpiContext::new_with_signer(
            ctx.accounts.conditional_vault_program.to_account_info(),
            conditional_vault::cpi::accounts::InteractWithVault {
                question: ctx.accounts.question.to_account_info(),
                vault: ctx.accounts.base_vault.to_account_info(),
                vault_underlying_token_account: ctx
                    .accounts
                    .base_vault_underlying_token_account
                    .to_account_info(),
                authority: ctx.accounts.futarchy_amm.to_account_info(),
                user_underlying_token_account: ctx.accounts.amm_base_vault.to_account_info(),
                event_authority: ctx.accounts.vault_event_authority.to_account_info(),
                program: ctx.accounts.conditional_vault_program.to_account_info(),
                token_program: ctx.accounts.token_program.to_account_info(),
            },
            signer,
        )
        .with_remaining_accounts(vec![
            ctx.accounts.fail_base_mint.to_account_info(),
            ctx.accounts.pass_base_mint.to_account_info(),
            ctx.accounts.amm_fail_base_vault.to_account_info(),
            ctx.accounts.amm_pass_base_vault.to_account_info(),
        ]);

        match swap_type {
            SwapType::Buy => {
                let amount_to_split = output_amount.saturating_sub(amm_output_account.amount);
                if amount_to_split > 0 {
                    conditional_vault::cpi::split_tokens(
                        base_cpi_context,
                        amount_to_split,
                    )?
                }
            }
            SwapType::Sell => {
                let base_mergeable = std::cmp::min(
                    ctx.accounts.amm_fail_base_vault.amount,
                    ctx.accounts.amm_pass_base_vault.amount,
                );

                if base_mergeable > 0 {
                    conditional_vault::cpi::merge_tokens(base_cpi_context, base_mergeable)?
                }
            }
        }

        token::transfer(
            CpiContext::new_with_signer(
                ctx.accounts.token_program.to_account_info(),
                token::Transfer {
                    from: amm_output_account.to_account_info(),
                    to: ctx.accounts.user_output_account.to_account_info(),
                    authority: ctx.accounts.futarchy_amm.to_account_info(),
                },
                signer,
            ),
            output_amount,
        )?;

        // ctx.accounts.amm_base_vault.reload()?;
        // ctx.accounts.amm_quote_vault.reload()?;
        // ctx.accounts.amm_pass_base_vault.reload()?;
        // ctx.accounts.amm_pass_quote_vault.reload()?;
        // ctx.accounts.amm_fail_base_vault.reload()?;
        // ctx.accounts.amm_fail_quote_vault.reload()?;

        // msg!("fail base: {}", ctx.accounts.amm_fail_base_vault.amount);
        // msg!("fail quote: {}", ctx.accounts.amm_fail_quote_vault.amount);
        // msg!("pass base: {}", ctx.accounts.amm_pass_base_vault.amount);
        // msg!("pass quote: {}", ctx.accounts.amm_pass_quote_vault.amount);
        // msg!("base: {}", ctx.accounts.amm_base_vault.amount);
        // msg!("quote: {}", ctx.accounts.amm_quote_vault.amount);

        Ok(())
    }
}

File: fuzz/futarchy/instructions/conditional_swap.py
"""Snapshot-checked wrapper for Futarchy ``conditional_swap``."""

from __future__ import annotations

from wake_sol import Account, AnchorError, random, svm

from .base import InstructionWrapper
from ..constants import (
    CONDITIONAL_VAULT_PROGRAM_ID,
    FUTARCHY_PROGRAM_ID,
    TOKEN_SCALE,
    TWAP_UPDATE_INTERVAL_SECONDS,
)
from ..pytypes.futarchy import (
    ConditionalSwapParams,
    Futarchy as FutarchyProgram,
    Market,
    PoolState,
    ProposalAction,
    SwapType,
)
from ..utils.swaps import assert_swap_transition
from ..utils.state import ProposalAccounts
from ..utils.tokens import token_balance


class ConditionalSwapInstruction(InstructionWrapper):
    """Check conditional user transfers and all three reserve products."""

    @staticmethod
    def build_params(
        market: Market,
        direction: SwapType,
        amount: int,
        minimum_output: int = 0,
    ) -> ConditionalSwapParams:
        return ConditionalSwapParams(
            market=market,
            swapType=direction,
            inputAmount=amount,
            minOutputAmount=minimum_output,
        )

    @staticmethod
    def user_accounts(
        proposal: ProposalAccounts,
        trader: Account,
        market: Market,
        direction: SwapType,
    ) -> tuple[Account, Account]:
        pass_market = market == Market.Pass
        base_mint = proposal.pass_base_mint if pass_market else proposal.fail_base_mint
        quote_mint = (
            proposal.pass_quote_mint if pass_market else proposal.fail_quote_mint
        )
        base = proposal.conditional_accounts[(trader.pubkey, base_mint.pubkey)]
        quote = proposal.conditional_accounts[(trader.pubkey, quote_mint.pubkey)]
        return (quote, base) if direction == SwapType.Buy else (base, quote)

    def build_instruction(
        self,
        proposal: ProposalAccounts,
        trader: Account,
        input_account: Account,
        output_account: Account,
        params: ConditionalSwapParams,
    ):
        context = self.context
        context.market_support.ensure_launch_accounts(proposal)
        return FutarchyProgram.conditionalSwap(
            params,
            dao=context.dao,
            ammBaseVault=context.amm_base_vault,
            ammQuoteVault=context.amm_quote_vault,
            proposal=proposal.proposal,
            ammPassBaseVault=proposal.amm_pass_base_vault,
            ammPassQuoteVault=proposal.amm_pass_quote_vault,
            ammFailBaseVault=proposal.amm_fail_base_vault,
            ammFailQuoteVault=proposal.amm_fail_quote_vault,
            trader=trader,
            userInputAccount=input_account,
            userOutputAccount=output_account,
            baseVault=proposal.base_vault,
            baseVaultUnderlyingTokenAccount=proposal.base_vault_underlying,
            quoteVault=proposal.quote_vault,
            quoteVaultUnderlyingTokenAccount=proposal.quote_vault_underlying,
            passBaseMint=proposal.pass_base_mint,
            failBaseMint=proposal.fail_base_mint,
            passQuoteMint=proposal.pass_quote_mint,
            failQuoteMint=proposal.fail_quote_mint,
            conditionalVaultProgram=CONDITIONAL_VAULT_PROGRAM_ID,
            vaultEventAuthority=context.vault_event_authority,
            question=proposal.question,
            eventAuthority=context.event_authority,
            program=FUTARCHY_PROGRAM_ID,
        )

    def can_happy(self) -> bool:
        return not self.context.is_liquidated() and bool(self.available_trades())

    def available_trades(
        self,
    ) -> list[tuple[ProposalAccounts, Account, Market, SwapType]]:
        """Find trades that can obtain a positive conditional-token input."""
        context = self.context
        result: list[tuple[ProposalAccounts, Account, Market, SwapType]] = []
        for proposal in context.pending_proposals():
            action = context.proposal_state(proposal).action
            choices = (
                ((Market.Pass, SwapType.Buy), (Market.Fail, SwapType.Sell))
                if isinstance(action, ProposalAction.HostileLiquidate)
                else tuple(
                    (market, direction)
                    for market in (Market.Pass, Market.Fail)
                    for direction in (SwapType.Buy, SwapType.Sell)
                )
            )
            for trader in context.actors:
                for market, direction in choices:
                    pass_market = market == Market.Pass
                    mint = (
                        proposal.pass_quote_mint
                        if pass_market and direction == SwapType.Buy
                        else proposal.fail_quote_mint
                        if direction == SwapType.Buy
                        else proposal.pass_base_mint
                        if pass_market
                        else proposal.fail_base_mint
                    )
                    conditional = proposal.conditional_accounts.get(
                        (trader.pubkey, mint.pubkey)
                    )
                    underlying = (
                        context.quote_atas_by_owner[trader.pubkey]
                        if direction == SwapType.Buy
                        else context.base_atas_by_owner[trader.pubkey]
                    )
                    if (
                        conditional is not None
                        and conditional.exists
                        and token_balance(conditional) > 0
                    ) or token_balance(underlying) > 0:
                        result.append((proposal, trader, market, direction))
        return result

    def prepare_positive_trade(
        self,
    ) -> tuple[
        ProposalAccounts,
        Account,
        Market,
        SwapType,
        Account,
        Account,
    ]:
        """Select and fund one trade, then require a positive input balance."""
        proposal, trader, market, direction = random.choice(
            self.available_trades()
        )
        self.context.market_support.ensure_trader_tokens(proposal, trader)
        input_account, output_account = self.user_accounts(
            proposal, trader, market, direction
        )
        assert token_balance(input_account) > 0
        return (
            proposal,
            trader,
            market,
            direction,
            input_account,
            output_account,
        )

    def happy(self) -> None:
        """Trade pass/fail tokens and verify the observed output transition."""
        context = self.context
        (
            proposal,
            trader,
            market,
            direction,
            input_account,
            output_account,
        ) = self.prepare_positive_trade()
        action = context.proposal_state(proposal).action
        maximum = min(
            token_balance(input_account),
            (1_000 if isinstance(action, ProposalAction.HostileLiquidate) else 100)
            * TOKEN_SCALE,
        )
        amount = (
            maximum
            if isinstance(action, ProposalAction.HostileLiquidate)
            else random.randint(1, maximum)
        )
        params = self.build_params(market, direction, amount)
        instruction = self.build_instruction(
            proposal, trader, input_account, output_account, params
        )
        if isinstance(action, ProposalAction.HostileLiquidate):
            svm.warp_to_timestamp(
                svm.clock.unix_timestamp + TWAP_UPDATE_INTERVAL_SECONDS
            )
        before_dao = context.dao_state()
        timestamp = svm.clock.unix_timestamp
        before = self.snapshot(
            instruction,
            extra_accounts=(context.dao, input_account, output_account),
        )
        trader.tx(self.compute_unit_limit(), instruction)
        after_dao = context.dao_state()
        assert token_balance(input_account) == (
            before.account(input_account).token_balance() - amount
        )
        output = token_balance(output_account) - before.account(
            output_account
        ).token_balance()
        assert output >= params.minOutputAmount
        assert isinstance(before_dao.amm.state, PoolState.Futarchy)
        assert isinstance(after_dao.amm.state, PoolState.Futarchy)
        assert_swap_transition(
            before_dao, after_dao, market, direction, amount, output, timestamp
        )

    def can_unhappy(self) -> bool:
        return self.can_happy()

    def unhappy(self) -> None:
        """Reject Spot selection or positive impossible slippage atomically."""
        context = self.context
        (
            proposal,
            trader,
            market,
            direction,
            input_account,
            output_account,
        ) = self.prepare_positive_trade()
        amount = random.randint(
            1, min(token_balance(input_account), 100 * TOKEN_SCALE)
        )
        if random.choice((False, True)):
            params = self.build_params(Market.Spot, direction, amount)
            expected = AnchorError.RequireNeqViolated
        else:
            params = self.build_params(
                market, direction, amount, 2**64 - 1
            )
            expected = FutarchyProgram.SwapSlippageExceeded
        instruction = self.build_instruction(
            proposal, trader, input_account, output_account, params
        )
        self.assert_fails_atomically(
            trader,
            instruction,
            expected,
            before=(self.compute_unit_limit(),),
        )
